tla-io/src/main/scala/at/forsyte/apalache/io/quint/QuintTypeConverter.scala

```diff
@@ -36,24 +37,24 @@ private class QuintTypeConverter extends LazyLogging {
 
   import QuintType._
 
-  private def rowToTupleElems(row: Row): Seq[TlaType1] = {
-    // Since we have to concat lists here, which can be expensive, we use an
-    // accumulator to benefit from tail call optimization. This is purely a precaution.
-    def aux(r: Row, acc0: Seq[TlaType1]): Seq[TlaType1] = r match {
-      // TODO Update with support for row-based tuples: https://github.com/informalsystems/apalache/issues/1591
-      // Row variables in tuples are not currently supported. But we will proceed assuming that this is an
-      // over generalization, and that we can safely treat the tuple as a closed tuple type.
-      case Row.Var(_) =>
-        logger.debug(
-            s"Found open row variable in quint tuple with fields $row. Polymorphic tuples are not supported, but we will proceed assuming the tuple can be treated as a closed type.")
-        acc0
-      case Row.Nil() => acc0
+  private def rowToTupleT1(row: Row): TlaType1 = {
+    // `aux(r, acc)` is a `(fields, other)`
+    // where `fields` is a map from tuple indices to tuple types, and
+    // `other` is either `Some(var)` for an open row or `None` for a closed row.
+    //
+    // `acc` is used as an accumulator to enable tail recursion
+    def aux(r: Row, acc0: SortedMap[Int, TlaType1]): (SortedMap[Int, TlaType1], Option[VarT1]) = r match {
+      case Row.Nil()  => (acc0, None)
+      case Row.Var(v) => (acc0, Some(VarT1(getVarNo(v))))
       case Row.Cell(fields, other) =>
-        val acc1 = fields.map(f => convert(f.fieldType)) ++ acc0
+        val acc1 = acc0 ++ fields.map { f => f.fieldName.toInt + 1 -> convert(f.fieldType) }
```

Comment:
```suggestion
        val acc1 = acc0 ++ fields.map { f => (f.fieldName.toInt + 1) -> convert(f.fieldType) }
```

Otherwise, it looks like `1 -> ...`